src/VisualStudio/Core/Def/SampleToolWindow/SampleToolboxUserControl.xaml.cs

```diff
@@ -0,0 +1,460 @@
+        {
+            InitializeComponent();
+            InitializeIfNeeded(workspace, documentTrackingService, languageServiceBroker, threadingContext);
+        }
+
+        private Workspace? workspace { get; set; }
+        private DocumentId? lastDocumentId { get; set; }
+        private List<DocSymbol>? originalTree { get; set; }
+        private ITextSnapshot? snapshot { get; set; }
+        private IWpfTextView? textView { get; set; }
+
+        private void InitializeIfNeeded(Workspace workspace, IDocumentTrackingService documentTrackingService, ILanguageServiceBroker2 languageServiceBroker, IThreadingContext threadingContext)
+        {
+            this.workspace = workspace;
+            documentTrackingService.ActiveDocumentChanged += DocumentTrackingService_ActiveDocumentChanged;
+
+            this.textView = GetActiveTextView();
+            if (this.textView is not null)
+            {
+                this.textView.Caret.PositionChanged += FollowCursor;
+            }
+
+            searchBox.GotFocus += RemoveText;
+            searchBox.LostFocus += AddText;
+
+            async void DocumentTrackingService_ActiveDocumentChanged(object sender, DocumentId? documentId)
+            {
+                if (documentId == this.lastDocumentId)
+                {
+                    return;
+                }
+
+                this.lastDocumentId = documentId;
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                if (document is null)
+                {
+                    symbolTree.ItemsSource = new List<DocSymbol>();
+                    return;
+                }
+
+                document.TryGetText(out var text);
+                if (text is null)
+                {
+                    symbolTree.ItemsSource = new List<DocSymbol>();
+                    return;
+                }
+
+                var textBuffer = text.Container.GetTextBuffer();
+                this.snapshot = textBuffer.CurrentSnapshot;
+                var isCorrectType = textBuffer.ContentType.IsOfType(ContentTypeNames.RoslynContentType);
+
+                if (!isCorrectType)
+                {
+                    symbolTree.ItemsSource = new List<DocSymbol>();
+                    return;
+                }
+
+                var parameterFactory = new DocumentSymbolParams()
+                {
+                    TextDocument = new TextDocumentIdentifier()
+                    {
+                        Uri = document.GetURI()
+                    }
+                };
+
+                var serializedParams = JToken.FromObject(parameterFactory);
+                JToken ParameterFactory(ITextSnapshot _)
+                {
+                    return serializedParams;
+                }
+
+                // make LSP request
```

Comment:
💡My understanding is that we are doing two things: making ah LSP request and updating the UI. Perhaps we could extract the LSP request logic into its own method so that all the logic here is to just update the UI.